lib/msgraph/paginated.go

```diff
@@ -144,3 +151,66 @@ func (c *Client) IterateGroupMembers(ctx context.Context, groupID string, f func
 	}
 	return trace.Wrap(itErr)
 }
+
+const (
+	graphNamespaceGroups         = "microsoft.graph.group"
+	graphNamespaceDirectoryRoles = "microsoft.graph.directoryRole"
+)
+
+const (
+	securityGroupsFilter = `mailEnabled eq false and securityEnabled eq true`
+)
+
+// IterateUsersTransitiveMemberOf lists groups that the user is a member of
+// through a direct or nested group membership.
+// This method calls user's transitiveMemberOf endpoint https://learn.microsoft.com/en-us/graph/api/user-list-transitivememberof?view=graph-rest-1.0&tabs=http.
+// Supported endpoints:
+// - All groups and directory roles: /v1.0/users/<user-id>/transitiveMemberOf
+// - Security groups: /v1.0/users/<user-id>/transitiveMemberOf/microsoft.graph.group?$filter=mailEnabled eq false and securityEnabled eq true
+// - Directory roles: /v1.0/users/<user-id>/transitiveMemberOf/microsoft.graph.directoryRole
+// Only group ID is extracted from the response, so the DirectoryObject struct is sufficient
+// to parse groups as well ass directory roles response.
+func (c *Client) IterateUsersTransitiveMemberOf(ctx context.Context, userID, groupType string, f func(*Group) bool) error {
+	// MS Graph expects $count query parameter and
+	// "ConsistencyLevel: eventual" header set when using
+	// advanced query parameter such as $filter.
+	// https://learn.microsoft.com/en-us/graph/aad-advanced-queries?tabs=http#legend
+	query := url.Values{
+		"$select": {"id"},
+		"$count":  {"true"},
+	}
+	header := map[string]string{
+		"ConsistencyLevel": "eventual",
+	}
+
+	endpoint := path.Join("users", userID, "transitiveMemberOf")
+	switch groupType {
+	case types.EntraIDAllGroups:
+		// default endpoint suffices.
+	case types.EntraIDDirectoryRoles:
+		endpoint = path.Join(endpoint, graphNamespaceDirectoryRoles)
+	case types.EntraIDSecurityGroups:
+		endpoint = path.Join(endpoint, graphNamespaceGroups)
+		query.Add("$filter", securityGroupsFilter)
+	default:
+		trace.BadParameter("unexpected group type %q received, expected types are %q", groupType, types.EntraIDGroupsTypes)
```

Comment:
is this suppose to return ?
```suggestion
		return trace.BadParameter("unexpected group type %q received, expected types are %q", groupType, types.EntraIDGroupsTypes)
```